### ISSUE TITLE
RandomSurvivalForest - predict_survival_function

### ISSUE THREAD

**Issue by felipe0216** [sebp/scikit-survival#193]
**Describe the bug**

1. I am trying to predict the survival function for my data using RandomSurvivalForest, although the class method works well, it doesn't retrieve the times for each of the steps in the survival function. Each list containing the survival function has a lenght equal or lower to the number of unique times in our "y", hence we can't deduct to what point in time each steps belongs to.

2. Additionally, if you follow the example given in the documentation of RandomSurvivalForest, you will get the following error:

```python
from sksurv.datasets import load_whas500
X, y = load_whas500()
times = sorted(np.unique(y["lenfol"])) 
n_times = len(times) 
# n_times =  395

estimator = RandomSurvivalForest().fit(X, y)
surv_funcs = estimator.predict_survival_function(X.iloc[:5])

surv_funcs[0]
# array([0.9975    , 0.9975    , 0.9975    , 0.9975    , 0.9975    ,
#       0.9975    , 0.9975    , 0.995     , 0.98883333, 0.98883333,...

len(surv_funcs[0])
# 162

```

Additionally, if you follow the example given in the documentation of RandomSurvivalForest, you will get an error since the result of `predict_survival_function` is an 1D unlike the same function used in `CoxnetSurvivalAnalysis` or `CoxPHSurvivalAnalysis`. This is the error you get:

```python
from sksurv.datasets import load_whas500
X, y = load_whas500()
estimator = RandomSurvivalForest().fit(X, y)
surv_funcs = estimator.predict_survival_function(X.iloc[:5])
for fn in surv_funcs:
       plt.step(fn.x, fn(fn.x), where="post")

plt.ylim(0, 1)
plt.show()

AttributeError: 'numpy.ndarray' object has no attribute 'x'
```

**Comment by sebp** [comment]
You are probably using an old version of scikit-survival, where 
predict_survival_function was not returning an array of StepFunction 
instances by default (return_array=True). StepFunction instances can be 
evaluated at any time point and will return the corresponding survival 
probability, as done in the user guide.

Can you please post the version you are using?

**Comment by felipe0216** [comment]
Hello @sebp ,

I currently have the version 0.14.0. From what I can see the latest version is 0.15.1. Although I can't get hold of that version neither by using `pip install scikit-survival` or `conda install -c sebp scikit-survival`, if I tried getting the development version using `pip install git+https://github.com/sebp/scikit-survival.git` as suggested in the guideline, but couldn't get it as it says `ERROR: Could not build wheels for scikit-survival which use PEP 517 and cannot be installed directly`

I am currently using Python 3.8.3, setuptools-56.0.0 wheel-0.36.2, pip 21.0.1

**Comment by sebp** [comment]
The latest released version is 0.15.0, which is available via conda and 
pip. Version 0.15.1 is the current (unreleased) development version, 
which can only be installed from source.

**Comment by felipe0216** [comment]
Hi sebp,

Thanks, I got that, my problem is now with the error `Could not build wheels for scikit-survival which use PEP 517 and cannot be installed directly`, I have read the guidelines and I have visited many forums to try find a solution for this. I have downloaded the `Visual Studio Build Tools 2019` and marked all the C/C++ options and Windows SDK for the latest version and some slightly older ones just in case. Also, I have installed the `Visual Studio Community 2019` since the above didn't work and did the same with all the `Desktop development with C++` as well as the `Universal Windows Platform development`. At this point I really don't know what the problem is, if you have any suggestions, please let me know.

Thanks !

**Comment by sebp** [comment]
Can you please run pip with the "-v" option to get detailed output. It 
should provide more detail at which step it fails. From past experience, 
it often is osqp, which requires cmake for compilation.

**Comment by felipe0216** [comment]
Hi @sebp ,

Thanks, I have run the command and this is the output. I highly appreciate the help.

In my untrained eye for these things, it seems that it could be that it is pointing at the default directory (AppData\Local\Temp\1\) rather than the conda directory ?

I have tried to trimmed the outputs as much as possible, cutting the irrelavant bits

```
Found link https://files.pythonhosted.org/packages/f4/7b/d415b0c89babf23dcd8ee631015f043e2d76795edd9c7359d6e63257464b/scikit-learn-0.24.1.tar.gz#sha256=a0334a1802e64d656022c3bfab56a73fbd6bf4b1298343f3688af2151810bbdf (from https://pypi.org/simple/scikit-learn/) (requires-python:>=3.6), version: 0.24.1
Found link https://files.pythonhosted.org/packages/d6/04/ea17391926a03db3dd7c41d2090b7f4394b3be4be55fef49d6a5a6f3d796/scikit_learn-0.23.2-cp38-cp38-win_amd64.whl#sha256=1b8a391de95f6285a2f9adffb7db0892718950954b7149a70c783dc848f104ea (from https://pypi.org/simple/scikit-learn/) (requires-python:>=3.6), version: 0.23.2   
  Found link https://files.pythonhosted.org/packages/0d/34/9f6b2cfe50ecf7feb3ea08f2e620f30cfb5084350c4c06beffc4efacd5e3/scikit-learn-0.24.dev0.tar.gz#sha256=a0b2eda09690baabbb2c62c9d2a96922abada2288210324084a57f50080ce80d (from https://pypi.org/simple/scikit-learn/) (requires-python:>=3.6), version: 0.24.dev0
Given no hashes to check 4 links for project 'scikit-learn': discarding no candidates
Collecting scikit-learn<0.25,>=0.24.0
  Created temporary directory: C:\Users\felipe.araya\AppData\Local\Temp\1\pip-unpack-xvrr0nk1
  Getting credentials from keyring for files.pythonhosted.org
  Looking up "https://files.pythonhosted.org/packages/7b/57/502e640e110e5ca5765036f832f88ed9ce529aa3e6854daa394389353b96/scikit_learn-0.24.1-cp38-cp38-win_amd64.whl" in the cache
  Current age based on date: 3278222
  Ignoring unknown cache-control directive: immutable
  Freshness lifetime from max-age: 365000000
  The response is "fresh", returning cached response
  365000000 > 3278222
  Using cached scikit_learn-0.24.1-cp38-cp38-win_amd64.whl (6.9 MB)
  Added scikit-learn<0.25,>=0.24.0 from https://files.pythonhosted.org/packages/7b/57/502e640e110e5ca5765036f832f88ed9ce529aa3e6854daa394389353b96/scikit_learn-0.24.1-cp38-cp38-win_amd64.whl#sha256=ed9d65594948678827f4ff0e7ae23344e2f2b4cabbca057ccaed3118fdc392ca (from scikit-survival) to build tracker 'C:\\Users\\felipe.araya\\AppData\\Local\\Temp\\1\\pip-req-tracker-g3sz7bi3'
  Removed scikit-learn<0.25,>=0.24.0 from https://files.pythonhosted.org/packages/7b/57/502e640e110e5ca5765036f832f88ed9ce529aa3e6854daa394389353b96/scikit_learn-0.24.1-cp38-cp38-win_amd64.whl#sha256=ed9d65594948678827f4ff0e7ae23344e2f2b4cabbca057ccaed3118fdc392ca (from scikit-survival) from build tracker 'C:\\Users\\felipe.araya\\AppData\\Local\\Temp\\1\\pip-req-tracker-g3sz7bi3'
Requirement already satisfied: joblib in c:\users\felipe.araya\anaconda3\lib\site-packages (from scikit-survival) (0.16.0)
Requirement already satisfied: scipy!=1.3.0,>=1.0 in c:\users\felipe.araya\anaconda3\lib\site-packages (from scikit-survival) (1.5.0)
Requirement already satisfied: future in c:\users\felipe.araya\anaconda3\lib\site-packages (from osqp!=0.6.0,!=0.6.1->scikit-survival) (0.18.2)
Requirement already satisfied: python-dateutil>=2.6.1 in c:\users\felipe.araya\anaconda3\lib\site-packages (from pandas>=0.25->scikit-survival) (2.8.1)
Requirement already satisfied: pytz>=2017.2 in c:\users\felipe.araya\anaconda3\lib\site-packages (from pandas>=0.25->scikit-survival) (2020.1)
Requirement already satisfied: six>=1.5 in c:\users\felipe.araya\anaconda3\lib\site-packages (from python-dateutil>=2.6.1->pandas>=0.25->scikit-survival) (1.15.0)Requirement already satisfied: threadpoolctl>=2.0.0 in c:\users\felipe.araya\anaconda3\lib\site-packages (from scikit-learn<0.25,>=0.24.0->scikit-survival) (2.1.0)
Created temporary directory: C:\Users\felipe.araya\AppData\Local\Temp\1\pip-unpack-vahral5s
Building wheels for collected packages: scikit-survival
  Created temporary directory: C:\Users\felipe.araya\AppData\Local\Temp\1\pip-wheel-1irbesjo
  Destination directory: C:\Users\felipe.araya\AppData\Local\Temp\1\pip-wheel-1irbesjo
  Running command 'c:\users\felipe.araya\anaconda3\python.exe' 'c:\users\felipe.araya\anaconda3\lib\site-packages\pip\_vendor\pep517\_in_process.py' build_wheel 'C:\Users\FELIPE~1.ARA\AppData\Local\Temp\1\tmpeye2mg2b'
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.8
  creating build\lib.win-amd64-3.8\sksurv
  copying sksurv\base.py -> build\lib.win-amd64-3.8\sksurv
  copying sksurv\column.py -> build\lib.win-amd64-3.8\sksurv
  copying sksurv\datasets\base.py -> build\lib.win-amd64-3.8\sksurv\datasets
  copying sksurv\datasets\__init__.py -> build\lib.win-amd64-3.8\sksurv\datasets
  creating build\lib.win-amd64-3.8\sksurv\ensemble
  copying sksurv\ensemble\boosting.py -> build\lib.win-amd64-3.8\sksurv\ensemble  
copying sksurv\ensemble\survival_loss.py -> build\lib.win-amd64-3.8\sksurv\ensemble
  copying sksurv\ensemble\__init__.py -> build\lib.win-amd64-3.8\sksurv\ensemble
  creating build\lib.win-amd64-3.8\sksurv\io
  copying sksurv\io\arffread.py -> build\lib.win-amd64-3.8\sksurv\io
copying sksurv\tree\__init__.py -> build\lib.win-amd64-3.8\sksurv\tree
  running egg_info
  writing scikit_survival.egg-info\PKG-INFO
  writing dependency_links to scikit_survival.egg-info\dependency_links.txt
  writing requirements to scikit_survival.egg-info\requires.txt
  writing top-level names to scikit_survival.egg-info\top_level.txt
  adding license file 'COPYING' (matched pattern 'COPYING*')
  reading manifest file 'scikit_survival.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  no previously-included directories found matching 'doc\_build'
  no previously-included directories found matching 'doc\generated'
  no previously-included directories found matching '.github'
  no previously-included directories found matching 'ci'
  warning: no previously-included files found matching 'appveyor.yml'
  warning: no previously-included files found matching '.codecov.yml'
  warning: no previously-included files found matching '.coveragerc'
  warning: no previously-included files found matching '.gitmodules'
  warning: no previously-included files found matching '.zenodo.json'
  warning: no previously-included files matching '__pycache__' found anywhere in distribution
  warning: no previously-included files matching '.ipynb_checkpoints' found anywhere in distribution
  warning: no previously-included files matching '.gitignore' found anywhere in distribution
  warning: no previously-included files matching '*.pyc' found anywhere in distribution
  warning: no previously-included files matching '*.bak' found anywhere in distribution
  warning: no previously-included files matching '*.swp' found anywhere in distribution
  warning: no previously-included files matching '*~' found anywhere in distribution
  writing manifest file 'scikit_survival.egg-info\SOURCES.txt'
  copying sksurv\bintrees\_binarytrees.cpp -> build\lib.win-amd64-3.8\sksurv\bintrees
creating build\lib.win-amd64-3.8\sksurv\datasets\data
  copying sksurv\datasets\data\GBSG2.arff -> build\lib.win-amd64-3.8\sksurv\datasets\data
  copying sksurv\datasets\data\whas500.arff -> build\lib.win-amd64-3.8\sksurv\datasets\data
  copying sksurv\ensemble\_coxph_loss.pyx -> build\lib.win-amd64-3.8\sksurv\ensemble
  copying sksurv\kernels\_clinical_kernel.pyx -> build\lib.win-amd64-3.8\sksurv\kernels
  copying sksurv\linear_model\_coxnet.pyx -> build\lib.win-amd64-3.8\sksurv\linear_model
  copying sksurv\linear_model\src\eigen\Eigen\src\plugins\MatrixCwiseUnaryOps.h -> build\lib.win-amd64-3.8\sksurv\linear_model\src\eigen\Eigen\src\plugins        
  copying sksurv\svm\_minlip.pyx -> build\lib.win-amd64-3.8\sksurv\svm
  copying sksurv\svm\_prsvm.pyx -> build\lib.win-amd64-3.8\sksurv\svm
  copying sksurv\tree\_criterion.pyx -> build\lib.win-amd64-3.8\sksurv\tree
  running build_ext
  building 'sksurv.bintrees._binarytrees' extension
  creating build\temp.win-amd64-3.8
  creating build\temp.win-amd64-3.8\Release
  creating build\temp.win-amd64-3.8\Release\sksurv
  creating build\temp.win-amd64-3.8\Release\sksurv\bintrees
  C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\bin\HostX86\x64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Isksurv/bintrees -IC:\Users\felipe.araya\AppData\Local\Temp\1\pip-build-env-k9__x1by\overlay\Lib\site-packages\numpy\core\include -Ic:\users\felipe.araya\anaconda3\include -Ic:\users\felipe.araya\anaconda3\include "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.28.29910\include" /EHsc /Tpsksurv/bintrees/_binarytrees.cpp /Fobuild\temp.win-amd64-3.8\Release\sksurv/bintrees/_binarytrees.obj
  _binarytrees.cpp
  c:\users\felipe.araya\anaconda3\include\pyconfig.h(59): fatal error C1083: Cannot open include file: 'io.h': No such file or directory
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community\\VC\\Tools\\MSVC\\14.28.29910\\bin\\HostX86\\x64\\cl.exe' failed with exit status 2
  Building wheel for scikit-survival (PEP 517) ... error
  ERROR: Failed building wheel for scikit-survival
Failed to build scikit-survival
ERROR: Could not build wheels for scikit-survival which use PEP 517 and cannot be installed directly
Exception information:
Traceback (most recent call last):
  File "c:\users\felipe.araya\anaconda3\lib\site-packages\pip\_internal\cli\base_command.py", line 189, in _main
    status = self.run(options, args)
  File "c:\users\felipe.araya\anaconda3\lib\site-packages\pip\_internal\cli\req_command.py", line 178, in wrapper
    return func(self, options, args)
  File "c:\users\felipe.araya\anaconda3\lib\site-packages\pip\_internal\commands\install.py", line 358, in run
    raise InstallationError(
pip._internal.exceptions.InstallationError: Could not build wheels for scikit-survival which use PEP 517 and cannot be installed directly
Removed build tracker: 'C:\\Users\\felipe.araya\\AppData\\Local\\Temp\\1\\pip-req-tracker-g3sz7bi3'

```

**Comment by sebp** [comment]
It could be that the Windows SDK is not installed, at least that's what 
the following StackOverflow posts suggest:

- 
https://stackoverflow.com/questions/40018405/cannot-open-include-file-io-h-no-such-file-or-directory

- 
https://stackoverflow.com/questions/45904828/io-h-no-such-file-or-directory-during-pip-install-netifaces

As an alternative, you can try downloading and installing the wheel 
that's AppVeyor CI builds. See the link bewloe for latest commit from 
master: 
https://ci.appveyor.com/project/sebp/scikit-survival/build/job/o4ic1ss88ac9b9xj/artifacts

**Comment by felipe0216** [comment]
Hi @sebp ,

I finally managed to update to get the 0.15.0 version. Just for anyone who sees this in the future, even though the solutions that @sebp posted here seemed to work for most people, they didn't quite work for me. What ended up working for me was reinstalling everything from scratch (Anaconda, VS Code, VS Built Tools) and then doing `conda update conda`, `conda update anaconda`, `conda update conda`. By doing that I could solve one of the many errors I had with `setuptools`. After that I re-installed `scikit-learn`. Also, I had to ask my organisation to allow me some user permissions because I came across some issues with restrictions imposed by the company, so also keep that in mind. 

Thank you @sebp for you help. Awesome library and awesome work you have done here
